File: Data Structures & Algorithms/is-anagram/submission-1.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        // first guard
        if (s.length() != t.length()) {
            return false;
        }

        std:: map<char, int> m1;
        std:: map<char, int> m2;

        // same length
        for (int i = 0; i < s.length(); ++i) {
            m1[s[i]]++;
            m2[t[i]]++;
        }

        return m1 == m2;
    }
};

File: Data Structures & Algorithms/intersection-of-two-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        
        # input:
        # headA
        # headB
        # ouput:
        # return the node intersects
        # no intersection -> null

        # constraints:
        # m, n >= 1
        # Node.val >= 1
        # no circles

        # edge cases:
        # Same head -> return that
        
        dup : set = set()
        traverse1 = headA
        traverse2 = headB

        while traverse1:
            dup.add(traverse1)
            traverse1 = traverse1.next

        while traverse2:
            if traverse2 in dup:
                return traverse2
            
            traverse2 = traverse2.next

        return None